### DISCUSSION TITLE
Continuous clamp (e.g. to `arccos`)

### DISCUSSION THREAD

**Discussion by rkruegs123** [jax-ml/jax#12295] [Q&A] [answered]
I am in the business of computing some inverse cosines with `arccos`. However, due to numerical issues, my execution sometimes results in the scenario where I try to take the arccos of a value outside of `[-1, 1]` by some epsilon, e.g. `jnp.arccos(1.0001)`. Of course this behavior is undefined, so I want to clamp. I've been using the following function to clamp:
```
def clamp(x, lo=-1.0, hi=1.0):
    min_ = jnp.where(x >= hi, hi, x)
    max_ = jnp.where(min_ <= lo, lo, min_)
    return max_
```
and using it like `jnp.arccos(clamp(x))`. Does anyone have any better recommendations for clamping, or ensuring that the domain for `arccos` is valid? Note that I need a continuous implementation of clamping as I am taking gradients.

Thank you!

**Comment by jakevdp** [comment, marked as answer]
I would probably use JAX's `clamp` primitive instead:
```python
>>> x = 1.01
>>> print(lax.clamp(0.0, x, 1.0))                                                                                                                                                                                                                                     
1.0
```
This has an appropriate autodiff rule defined, although by nature clamping of values may result in some strange gradients depending on how it is used in your computation.

**Reply by rkruegs123** [reply]
Thank you, Jake. I used `jnp.clamp` but not the primitive -- I'll use this and follow up. I'm hoping it doesn't lead to too weird of behavior as it's only very minor epsilons that we have to clamp.

Thanks again for the quick reply.

**Reply by jakevdp** [reply]
Well, regardless of what method you use, I suspect with gradients you'll run into the issues mentioned here: https://jax.readthedocs.io/en/latest/faq.html#why-are-gradients-zero-for-functions-based-on-sort-order

No matter how you phrase it, the gradients of a clip/clamp function are going to be zero for values larger than the clip range, and this is not a bug: zero is the correct analytic gradient in this case.

**Reply by jakevdp** [reply]
Probably what you want is to use something like [`smoothstep`](https://en.wikipedia.org/wiki/Smoothstep) in place of `clamp`; this trades off accuracy near `x=0` and `x=1` for smoothness/differentiability of the clamp-like operation.

**Reply by rkruegs123** [reply]
I see... that makes a lot of sense regarding the gradients and when you clamp outside the range...

I saw `smoothstep` and, though it is minor, don't love the cost in accuracy. What do you think about my temporary solution using `jnp.where`? It seems to be both continuous and accurate, but I am worried that the gradients won't be as "clean." For example, why would `jnp.clamp` not be implemented this way?

**Reply by jakevdp** [reply]
The two approaches seem to have equivalent and correct gradient behavior... what do you mean when you say "clean" in this context?
```python
import jax.numpy as jnp
from jax import lax
import jax

def clamp(x, lo=-1.0, hi=1.0):
    min_ = jnp.where(x >= hi, hi, x)
    max_ = jnp.where(min_ <= lo, lo, min_)
    return max_

def lax_clamp(x, lo=-1.0, hi=1.0):
    return lax.clamp(lo, x, hi)

x = jnp.array([-1.1, -1.0, -0.1, 0.0, 0.1, 1.0, 1.1])

print(jax.vmap(jax.grad(clamp))(x))
# [0. 0. 1. 1. 1. 0. 0.]

print(jax.vmap(jax.grad(lax_clamp))(x))
# [0. 0. 1. 1. 1. 0. 0.]
```

**Reply by rkruegs123** [reply]
Ah, I did not know that was the gradient behavior of `jnp.where`! But it makes sense...

Thank you for being so responsive and helpful. Closing this, and thank you again!
